typeinfo/typeinfo.go

```diff
@@ -0,0 +1,137 @@
+package typeinfo
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+)
+
+var cmutex sync.RWMutex
+var cache = make(map[string]Info)
+
+// Return the Info of a type name passed as parameter
+// Returns error if the type has not been reflected yet.
+func GetInfoFromName(typeName string) (Info, error) {
+	cmutex.RLock()
+	info, found := cache[typeName]
+	cmutex.RUnlock()
+	if found {
+		return info, nil
+	} else {
+		return info, fmt.Errorf("unknown type")
+	}
+}
+
+// Reflect will return the Info of a given type,
+// generating and caching as required.
+func GetTypeInfo(value any) (Info, error) {
+	if value == (any)(nil) {
+		return Info{}, fmt.Errorf("Can not reflect nil value")
+	}
+
+	v := reflect.ValueOf(value)
+	v = reflect.Indirect(v)
+
+	tname := v.Type().Name()
+
+	cmutex.RLock()
+	info, found := cache[tname]
+	cmutex.RUnlock()
+	if found {
+		return info, nil
+	}
+
+	info, err := generate(v)
+	if err != nil {
+		return Info{}, err
+	}
+
+	// Do not cache for "M" types
+	if !(info.Type.Kind() == reflect.Map && info.Type.Name() == "M") {
```

Comment:
The `sqlair.M` was in the [original specification](https://docs.google.com/document/d/1FbRZisjK0z7shy8m1cHzzEMDpxGwsK-54xDpf9HY4Xo/edit#) and we have kept it that way. It was also mentioned in the sprint during a meeting with the juju team.

Reply:
Again, what's the plan about `M` handling here?

Reply:
Sorry, I'm missing something here. We are caching by type. So if we have something like:
```
       var mymap M
       mymap = make(M)
       mymap["foo"] = 7 
       mymap["bar"] = "baz"
...
...
       var mysecondmap M
       mysecondmap = make(M)
       info, err := GetTypeInfo(mysecondmap)
       assert.Nil(t, err)
       assert.Len(t, info.TagToField, 0)
```
If we retrieve data from the cache, the last `assert.Len(t, info.TagToField, 0)` will fail. It will report length = 2, instead of the expected 0.

Reply:
What tags and field names will be cached for `M`? What tags and field names will be cached for other maps?

Reply:
The plan for `M` is not to cache anything. We don't support other maps (as in generic maps) AFAIK. There is no mention to that in the specification. Either `structs` or `sqlair.M` with `M` being of underlying type `map[string]any`. Is the counter example above ill formed?

Reply:
I changed it so we don't try to store key information whatsoever. I would not be necessary according to a [concept](https://github.com/fernape/sqlairexperiment/blob/minimal-reflection/typeinfo/info.go#L40) we've been working on.